src/nvme/ioctl.c

```diff
@@ -355,6 +355,130 @@ int nvme_get_log_page(int fd, __u32 xfer_len, struct nvme_get_log_args *args)
 	return 0;
 }
 
+static int read_ana_chunk(int fd, enum nvme_log_ana_lsp lsp, bool rae,
+			  __u8 *log, __u8 **read, __u8 *to_read, __u8 *log_end)
+{
+	if (to_read > log_end) {
+		errno = ENOSPC;
+		return -1;
+	}
+
+	while (*read < to_read) {
+		__u32 len = min(log_end - *read, NVME_LOG_PAGE_PDU_SIZE);
+		int ret = nvme_get_log_ana(fd, lsp, rae, *read - log, len, *read);
+
+		if (ret)
+			return ret;
+
+		*read += len;
+	}
+	return 0;
+}
+
+static int try_read_ana(int fd, enum nvme_log_ana_lsp lsp, bool rae,
+			struct nvme_ana_log *log, __u8 *log_end,
+			__u8 *read, __u8 **to_read, bool *may_retry)
+{
+	__u16 ngrps = le16_to_cpu(log->ngrps);
+
+	while (ngrps--) {
+		__u8 *group = *to_read;
+		int ret;
+		__le32 nnsids;
+
+		*to_read += sizeof(*log->descs);
+		ret = read_ana_chunk(fd, lsp, rae,
+				     (__u8 *)log, &read, *to_read, log_end);
+		if (ret) {
+			/*
+			 * If the provided buffer isn't long enough,
+			 * the log page may have changed while reading it
+			 * and the computed length was inaccurate.
+			 * Have the caller check chgcnt and retry.
+			 */
+			*may_retry = errno == ENOSPC;
+			return ret;
+		}
+
+		/*
+		 * struct nvme_ana_group_desc has 8-byte alignment
+		 * but the group pointer is only 4-byte aligned.
+		 * Don't dereference the misaligned pointer.
+		 */
+		memcpy(&nnsids,
+		       group + offsetof(struct nvme_ana_group_desc, nnsids),
+		       sizeof(nnsids));
+		*to_read += le32_to_cpu(nnsids) * sizeof(__le32);
+		ret = read_ana_chunk(fd, lsp, rae,
+				     (__u8 *)log, &read, *to_read, log_end);
+		if (ret) {
+			*may_retry = errno == ENOSPC;
+			return ret;
+		}
+	}
+
+	*may_retry = true;
+	return 0;
+}
+
+int nvme_get_ana_log_atomic(
+	int fd, bool rgo, bool rae, unsigned int max_retries,
+	struct nvme_ana_log *log, __u32 *len)
```

Comment:
Just a bikeshed question: the name 'atomic' is heavily overloaded and I understood it wrong initially. I was expecting that the this code will read everything in one go, which it isn't. It carefully makes sure it reads the data with a understanding of the data types/sizes/alignments within the log page. Maybe we can come up with something else as name? But I know naming is diffucult...

Reply:
I removed the line break before the start of the arguments, hopefully this is what you were asking for?

Reply:
aligment: all other functions in this file align the arguments with the open bracket. checkpatch is accepting different styles but doesn't report inconsistency. I don't care  too much about one or other style but I do care about consistency. 

naming: Whenever I read atomic I think in terms of cmpxchg. So we either do the whole operation or nothing. When I was reviewing it, it took me a while to figure out what you mean with atomic. The implementation makes sure it reads the individual parts of the log page in one go, respecting the type/size of each part. Sure, this can be considered 'atomic' but the whole function as such is not doing this as atomic operation. I got even more confused by 'chunk' function which I suppose should the the 'atomic' part? Anyway, I agree with Christoph and think the term atomic is heavily overloaded in it's meaning which can cause confusion.

Reply:
aligment (*sic*): I assume by "bracket" you mean the open parenthesis? I believe I have already made this change, can you please let me know if the updated review is to your aesthetic preferences?

naming: I don't mean to be pedantic here, but using a sequence number to ensure multiple reads are performed coherently certainly seems to qualify for the computer science definition of "atomic". "atomic" comes from a Greek root meaning roughly "indivisible". In computer science, it is widely used to describe any interaction between concurrent computations that is performed in an all-or-nothing way. The computation may involve multiple steps, in which case they are made to appear as a single unit. Typically this is done with some form of mutual exclusion or repeating a computation until the value it operates on doesn't change in the middle. cmpxchg is certainly an example of an atomic operation, and is the central building block for that "restartable computation" approach to perform more complex operations atomically. (Keep in mind that the processor's own implementation of cmpxchg requires multiple steps--read, compare, write--that have to be synchronized to make the operation appear atomic. So by your reasoning, a cmpxchg instruction wouldn't count as atomic either.) And in fact what we are doing here with the ANA change count is exactly a cmpxchg loop, except the cmpxchg is just an atomic read because there is no need to modify the change count:
```
1. Read change count
2. Read rest of ANA log page
3. Read change count again; if changed, goto 2
```
(If the whole log page fits in 4 KB, then only a single Get Log Page command is used, which is defined in the spec to be atomic. In that case the change count check is skipped.)
I am not wedded to "atomic", though. I have already suggested "untorn" as a possible alternative, and I'll suggest another: "coherent". Please let me know if you prefer one of those, or have your own suggestion. But "the term atomic...can cause confusion" is not really constructive feedback...

Reply:
Sorry for the late response.

Formatting: I know, it sometimes a PITA but such things matter none the less.

Naming: The point I wanted to make is that the context is important. Not everyone understands one word the same way. Anyway, I am fine as it is. So let's go with it.